core/includes/bootstrap.inc

```diff
@@ -854,21 +854,44 @@ function backdrop_settings_initialize() {
     require_once BACKDROP_ROOT . '/core/includes/drupal.inc';
   }
 
-  // Convert simplified database settings into the full databases array.
-  if (!empty($database) && $database !== 'mysql://user:pass@localhost/database_name') {
-    $database_parts = parse_url($database);
-    if (!$database_parts) {
-      trigger_error('The database setting could not be parsed. Please check the $database setting in settings.php.', E_USER_ERROR);
+  // Define default database setting values.
+  $default_db_string = 'mysql://user:pass@localhost/database_name';
+  $default_db_array = array(
+    'driver' => 'mysql',
+    'database' => 'database_name',
+    'username' => 'user',
+    'password' => 'pass',
+    'host' => 'localhost',
+    'prefix' => '',
+  );
+
+  // If database settings have been changed from default add to the full
+  // databases array.
+  if (!empty($database) && $database !== $default_db_string && $database !== $default_db_array) {
```

Comment:
`$database !== $default_db_array` expects the keys in the two arrays being compared to also be in the same order. I don't think that we should be that strict here - lets use `!=` instead.

I would also explicitly and separately check for `(is_string($database)` and `(is_array($database)`, which would allow to only define the `$default_db_string` and `$default_db_array` variables when needed (since they aren't being used anywhere else outside the conditional). In other words, instead of doing this:
```php
if (!empty($database) && $database !== $default_db_string && $database !== $default_db_array) {
  if (is_array($database)) {
    ...
  }
  else {
    ...
  }
}
```

I would prefer doing this:
```php
// Database connection settings have been provided in the form of a connection
// string.
if (!empty($database) && is_string($database)) {
  $default_db_string = ... ;
  if ($database !== $default_db_string) { ... }
}
// Database connection settings have been provided in the form of a simplified
// array.
if (!empty($database) && is_array($database)) {
  $default_db_array = ... ;
  if ($database != $default_db_array) { ... }
}
```

Reply:
Done